### DISCUSSION TITLE
Compiling OCaml to WebAssembly (Wasm)

### DISCUSSION THREAD

**Discussion by XantreDev** [ocaml/ocaml#12283] [Features]
Wasm GC is implemented in chrome: https://chromestatus.com/feature/6062715726462976 .

I would like to propose the development of a new target for OCaml that compiles to WebAssembly (Wasm). This initiative will be a strategic move forward for our language and community, with benefits that could have far-reaching effects for the future of web development.

1. **Fullstack Consistency:** With a Wasm target, we can use OCaml throughout the entire stack. This opens up opportunities for fullstack OCaml development, simplifying workflows, improving code reusability, and reducing the context switching that comes with using multiple languages.

2. **Expanding Use Cases:** The Wasm target will allow OCaml to be used in broader contexts. Wasm is becoming more prevalent on the web and in other environments like serverless computing. By enabling OCaml to compile to Wasm, we can help expand OCaml's presence in these growing areas.

3. **Strong Typing Advantage:** Our language's strong typing system can make web applications more robust and error-resistant, something that could provide a significant advantage over dynamically typed languages commonly used in web development.

4. **Performance:** The performance characteristics of Wasm, including its near-native execution speed, will allow applications written in OCaml to run quickly and efficiently in the browser.

5. **Interoperability:** Wasm is designed to interoperate smoothly with JavaScript. This makes it possible to incrementally adopt OCaml compiled to Wasm in existing projects without a complete rewrite.

6. **Growing Interest:** The functional programming paradigm has been gaining traction in the web development community. Capitalizing on this interest by offering a solid pathway for OCaml to Wasm compilation could attract more developers to our community.

To take advantage of these opportunities, we need to invest in the development of a robust OCaml to Wasm compiler. This is a significant undertaking, but the potential benefits for the OCaml community and the broader web development landscape are considerable. Your thoughts, feedback, and support on this initiative are much appreciated.

I can start investigating it if someone gives me good starting points

**Comment by gasche** [comment]
There already is an experimental prototype of wasm backend for OCaml, see [this presentation to the wasm GC group](https://github.com/WebAssembly/meetings/blob/fac1d6c7c174b774e325201a914cc99c0166de05/gc/2023/GC-01-10.md#presentation-ocaml-on-wasmgc-pierre-chambart-and-l%C3%A9o-andr%C3%A8s-60-min) by @zapashcanon and @chambart.

**Comment by redianthus** [comment]
Indeed. Currently we're working on formalizing and proving our compilation scheme.

We're also now confident about the fact that we'll be able to compile code written using jsoo to Wasm-only code (with the GC, exception and tail_call proposals available) without the need for users to change their code. We are currently implementing the feature in our runtime. In the (not so short) future we'll try to make C bindings available in Wasm by compiling the C bits with emscriptem and subverting the stubs.

We'll (hopefully :-)) present our work at the OCaml workshop if you're interested @XantreGodlike.

**Reply by XantreDev** [reply]
Sounds great. I'm hyped

**Comment by jhjourdan** [comment]
Do the recent versions of Wasm have *proper* support for tail call optimizations? I'm speaking about full TCO, not simply tail recursion. In my memory, this was the main blocking point!

**Reply by gasche** [reply]
No idea, but note that Javascript does not have proper TCO support either and people are still very interested in using js_of_ocaml (and occasionally ask to rewrite some functions that blow the stack on their inputs).

**Reply by redianthus** [reply]
@jhjourdan there's [a proposal for tail-calls](https://github.com/WebAssembly/tail-call/) - which has been ready for a long time. It is already implemented in Chrome and in many interpreters (such as [owi](https://github.com/OCamlPro/owi)). In order to become a part of the standard (phase 5), it needs to be implemented by two major browsers (phase 4) so we needed another one (i.e. Firefox or Safari). Safari recently implemented it and Firefox is having a hard time doing so as it requires rewriting a large part of SpiderMonkey...

It's been a long process but on the 17th of January [the proposal moved towards phase 4](https://github.com/WebAssembly/tail-call/issues/15#issuecomment-1385891079). You can have a look at [the full process](https://github.com/WebAssembly/meetings/blob/main/process/phases.md#4-standardize-the-feature-working-group) in order to see what's still to be done.

Note also that the main blocking point in order to get good performances was rather the [GC proposal](https://github.com/WebAssembly/gc/) which stabilized only recently. At the end of 2022, we implemented our experimental OCam to Wasm compiler in order to help the proposal move forward ([to convince people `i31ref`s were needed by some languages](https://github.com/WebAssembly/gc/issues/320)).

**Reply by XantreDev** [reply]
What do you think about this project? https://github.com/remixlabs/wasicaml
Seems it's tried to do the same thing

**Reply by redianthus** [reply]
It's not using Wasm-GC. Thus it is quite slow (slower than js_of_ocaml) and it makes interaction with the host harder. We gave a comparison [in the talk linked by Gabriel](https://github.com/WebAssembly/meetings/blob/fac1d6c7c174b774e325201a914cc99c0166de05/gc/2023/GC-01-10.md#presentation-ocaml-on-wasmgc-pierre-chambart-and-l%C3%A9o-andr%C3%A8s-60-min).

**Comment by omasanori** [comment]
[KC Sivaramakrishnan announced](https://discuss.ocaml.org/t/announcing-the-ocaml-wasm-organisation/12676) the [ocaml-wasm](https://github.com/ocaml-wasm) GitHub organization. Hooray!
